velox/functions/sparksql/DateTimeFunctions.cpp

```diff
@@ -0,0 +1,199 @@
+    return std::nullopt;
+  }
+
+  // year, month, day will be checked in utils::daysSinceEpochFromDate;
+  try {
+    auto daysSinceEpoch = util::daysSinceEpochFromDate(
+        yearVector->valueAt<int32_t>(row),
+        monthVector->valueAt<int32_t>(row),
+        dayVector->valueAt<int32_t>(row));
+    auto localMicros =
+        hour * util::kMicrosPerHour + minute * util::kMicrosPerMinute + micros;
+    return util::fromDatetime(daysSinceEpoch, localMicros);
+  } catch (const VeloxException& e) {
+    if (!e.isUserError()) {
+      throw;
+    }
+    return std::nullopt;
+  } catch (const std::exception&) {
+    throw;
+  }
+}
+
+void setTimestampOrNull(
+    int32_t row,
+    std::optional<Timestamp> timestamp,
+    int64_t tzID,
+    FlatVector<Timestamp>* result) {
+  if (timestamp.has_value()) {
+    (*timestamp).toGMT(tzID);
+    result->set(row, *timestamp);
+  } else {
+    result->setNull(row, true);
+  }
+}
+
+class MakeTimestampFunction : public exec::VectorFunction {
+ public:
+  MakeTimestampFunction(int64_t sessionTzID) : sessionTzID_(sessionTzID) {}
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, TIMESTAMP(), result);
+    auto* resultFlatVector = result->as<FlatVector<Timestamp>>();
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto year = decodedArgs.at(0);
+    auto month = decodedArgs.at(1);
+    auto day = decodedArgs.at(2);
+    auto hour = decodedArgs.at(3);
+    auto minute = decodedArgs.at(4);
+    auto micros = decodedArgs.at(5);
+
+    if (args.size() == 7) {
+      // If the timezone argument is specified, treat the input timestamp as the
+      // time in that timezone.
+      if (args[6]->isConstantEncoding()) {
+        auto constantTzID =
+            util::getTimeZoneID(args[6]
+                                    ->asUnchecked<ConstantVector<StringView>>()
+                                    ->valueAt(0)
+                                    .str());
+        rows.applyToSelected([&](vector_size_t row) {
+          auto timestamp = makeTimeStampFromDecodedArgs(
+              row, year, month, day, hour, minute, micros);
+          setTimestampOrNull(row, timestamp, constantTzID, resultFlatVector);
+        });
+      } else {
+        auto timeZone = decodedArgs.at(6);
+        rows.applyToSelected([&](vector_size_t row) {
+          auto timestamp = makeTimeStampFromDecodedArgs(
+              row, year, month, day, hour, minute, micros);
+          auto tzID =
+              util::getTimeZoneID(timeZone->valueAt<StringView>(row).str());
+          setTimestampOrNull(row, timestamp, tzID, resultFlatVector);
+        });
+      }
+    } else {
+      // Otherwise use session timezone. If session timezone is not specified,
+      // use default value 0(UTC timezone).
```

Comment:
I once noted Spark always has a session timezone in its config. By default, it's the one detected from OS. And we always let Gluten pass session timezone to Velox. So maybe, if session timezone is not found from config, we can simply throw an exception.